engine/render/src/render/buffer.cpp

```diff
@@ -0,0 +1,146 @@
+    {
+        DeleteRenderBuffer(buffer->m_Buffers.Back(), buffer->m_Type);
+        buffer->m_Buffers.Pop();
+    }
+
+    HBufferedRenderBuffer NewBufferedRenderBuffer(HRenderContext render_context, RenderBufferType type)
+    {
+        BufferedRenderBuffer* buffer = new BufferedRenderBuffer();
+        memset(buffer, 0, sizeof(BufferedRenderBuffer));
+        RewindBuffer(render_context, buffer);
+        buffer->m_Type = type;
+        return buffer;
+    }
+
+    void DeleteBufferedRenderBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
+    {
+        if (!buffer)
+            return;
+
+        for (int i = 0; i < buffer->m_Buffers.Size(); ++i)
+        {
+            DeleteRenderBuffer(buffer->m_Buffers[i], buffer->m_Type);
+        }
+        delete buffer;
+    }
+
+    HRenderBuffer AdvanceRenderBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
+    {
+        if (!buffer)
+            return 0;
+
+        if (buffer->m_Buffers.Empty())
+        {
+            buffer->m_Buffers.OffsetCapacity(1);
+            CreateAndPush(render_context, buffer);
+        }
+
+        if (!render_context->m_MultiBufferingRequired)
+        {
+            buffer->m_BufferIndex = 0;
+            assert(buffer->m_Buffers.Size() == 1);
+            return buffer->m_Buffers.Front();
+        }
+
+        buffer->m_BufferIndex++;
+        if (buffer->m_BufferIndex >= buffer->m_Buffers.Capacity())
+        {
+            buffer->m_Buffers.OffsetCapacity(1);
+            CreateAndPush(render_context, buffer);
+        }
+        return buffer->m_Buffers[buffer->m_BufferIndex];
+    }
+
+    void SetBufferData(HRenderContext render_context, HBufferedRenderBuffer buffer, uint32_t size, void* data, dmGraphics::BufferUsage buffer_usage)
+    {
+        switch(buffer->m_Type)
+        {
+            case RENDER_BUFFER_TYPE_VERTEX_BUFFER:
+            {
+                dmGraphics::HVertexBuffer vbuf = (dmGraphics::HVertexBuffer) buffer->m_Buffers[buffer->m_BufferIndex];
+                dmGraphics::SetVertexBufferData(vbuf, size, data, buffer_usage);
+            } break;
+            case RENDER_BUFFER_TYPE_INDEX_BUFFER:
+            {
+                dmGraphics::HIndexBuffer ibuf = (dmGraphics::HIndexBuffer) buffer->m_Buffers[buffer->m_BufferIndex];
+                dmGraphics::SetIndexBufferData(ibuf, size, data, buffer_usage);
+            } break;
+            default:assert(0);
+        }
+    }
+
+    void TrimBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
+    {
+        if (!buffer)
+            return;
+
+        uint32_t n = buffer->m_Buffers.Size();
+        uint32_t new_buffer_count = buffer->m_BufferIndex+1;
+
+        if (new_buffer_count == n)
+            return;
+
+        for (int i = 0; i < (n - new_buffer_count); ++i)
+        {
+            PopAndDelete(render_context, buffer);
+        }
+
+        buffer->m_Buffers.SetCapacity(new_buffer_count);
```

Comment:
While technically correct, will it lead to constant reallocations?
This buffer doesn't cost much, and if it grew to a certain size during one frame, it is very likely to be the case the next frame as well.

Also, I feel looping backwards is slightly harder to read at a glance.
Perhaps something like this:
```
void TrimBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
{
    uint32_t n = buffer->m_Buffers.Size();
    uint32_t new_buffer_count = buffer->m_BufferIndex+1;
    for (int i = new_buffer_count; i < n; ++i)
    {
        DeleteRenderBuffer(buffer->m_Buffers[i], buffer->m_Type);
    }
    buffer->m_Buffers.SetSize(new_buffer_count);
}
```